pkg/karmadactl/register.go

```diff
@@ -0,0 +1,981 @@
+	flags.Int32Var(&opts.CertExpirationSeconds, "cert-expiration-seconds", DefaultCertExpirationSeconds, "The expiration time of certificate.")
+	flags.BoolVar(&opts.DryRun, "dry-run", false, "Don't apply any changes; just output what would be done.")
+
+	return cmd
+}
+
+// CommandRegisterOption holds all command options.
+type CommandRegisterOption struct {
+	// KubeConfig holds the KUBECONFIG file path.
+	KubeConfig string
+
+	// Context is the name of the cluster context in KUBECONFIG file.
+	// Default value is the current-context.
+	Context string
+
+	// Namespace is the namespace that karmada-agent component deployed.
+	Namespace string
+
+	// ClusterNamespace holds the namespace name where the member cluster secrets are stored.
+	ClusterNamespace string
+
+	// ClusterName is the cluster's name that we are going to join with.
+	ClusterName string
+
+	// ClusterProvider is the cluster's provider.
+	ClusterProvider string
+
+	// ClusterRegion represents the region of the cluster locate in.
+	ClusterRegion string
+
+	// CACertPath is the path to the SSL certificate authority used to
+	// secure comunications between member cluster and karmada-control-plane.
+	// Defaults to "/etc/karmada/pki/ca.crt".
+	CACertPath string
+
+	// BootstrapToken is used to set the options for bootstrap token based discovery
+	BootstrapToken *BootstrapTokenDiscovery
+
+	// Timeout is the max discovery time
+	Timeout time.Duration
+
+	// CertExpirationSeconds define the expiration time of certificate
+	CertExpirationSeconds int32
+
+	// KarmadaSchedulerImage is the image of karmada agent.
+	KarmadaAgentImage string
+
+	// KarmadaAgentReplicas is the number of karmada agent.
+	KarmadaAgentReplicas int32
+
+	// DryRun tells if run the command in dry-run mode, without making any server requests.
+	DryRun bool
+
+	memberClusterEndpoint string
+	memberClusterClient   *kubeclient.Clientset
+}
+
+// Complete ensures that options are valid and marshals them if necessary.
+func (o *CommandRegisterOption) Complete(args []string) error {
+	// Get karmada apiserver endpoint from the command args.
+	if len(args) == 0 {
+		return fmt.Errorf("karmada apiserver endpoint is required")
+	}
+	o.BootstrapToken.APIServerEndpoint = args[0]
+
+	if o.KubeConfig == "" {
+		env := os.Getenv("KUBECONFIG")
+		if env != "" {
+			o.KubeConfig = env
+		} else {
+			o.KubeConfig = defaultKubeConfig
+		}
+	}
+
+	if !Exists(o.KubeConfig) {
+		return ErrEmptyConfig
+	}
+
+	restConfig, err := utils.RestConfig(o.Context, o.KubeConfig)
+	if err != nil {
+		return err
+	}
+
+	if len(o.ClusterName) == 0 {
+		configBytes, err := ioutil.ReadFile(o.KubeConfig)
+		if err != nil {
+			return fmt.Errorf("failed to read kubeconfig file %s, err: %w", o.KubeConfig, err)
+		}
+
+		config, err := clientcmd.Load(configBytes)
+		if err != nil {
+			return err
+		}
+
+		o.ClusterName = config.Contexts[config.CurrentContext].Cluster
```

Comment:
If users don't set `--cluster-name`, currently using default clusterName of `kubeconfig` of member cluster. But sometimes, most users install kubernetes in member cluster by default way, lead to the default clusterName is `kubernetes`.

```
[root@master68 lonelyCZ]# cat /root/.kube/config
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ......
    server: https://10.10.103.68:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: kubernetes-admin
  name: kubernetes-admin@kubernetes
current-context: kubernetes-admin@kubernetes
```

So how do we set default cluaterName if users don't set `--cluster-name`?

Reply:
> But sometimes, most users install kubernetes in member cluster by default way,

I don't know much about how users configured their clusters.
But I think if users are using multi-clusters, they should take good care of it. We can assume the default cluster name is well-configured. 
Yes, I mean, keeping the default is fine to me.

Reply:
Agree!